### PR TITLE
[WIP] Very rough initial draft of mmcif parser

### PR DESCRIPTION
**Pull request by a-r-j**
Hi Sebastian, thanks for a great project! I've relied on it a lot over the last few years. Hopefully I can give something back!

### Description
Adds (limited) support for mmcif files. I note on the issues it's not planned for the near term. Perhaps we can split up the work? Happy to help :)

### Related issues or pull requests
#80 #92 #93

### Pull Request Checklist

Noted, none of these have been done (yet!) as **PR is not ready for merge**.

- [x] Added a note about the modification or contribution to the `./docs/sources/CHANGELOG.md` file (if applicable)
- [x] Added appropriate unit test functions in the `./biopandas/*/tests` directories (if applicable)
- [x] Modify documentation in the corresponding Jupyter Notebook under `biopandas/docs/sources/` (if applicable)
- [x] Ran `PYTHONPATH='.' pytest ./biopandas -sv` and make sure that all unit tests pass (for small modifications, it might be sufficient to only run the specific test file, e.g., `PYTHONPATH='.' pytest ./biopandas/classifier/tests/test_stacking_cv_classifier.py -sv`)
- [ ] Checked for style issues by running `flake8 ./biopandas`

### CODE REVIEW

**Comment by rasbt** [comment]
Just skimmed through the code files. Wow, this is really cool stuff! Like I mentioned, I haven't worked with that particular file format (or even protein structures in general in the last few years), so I probably won't be much help here, unfortunately. However, I am happy to help looking at the big picture. Let me know in case you need anything.

**Comment by a-r-j** [comment]
Thanks! I've also not worked with them too much.

Wrt design, I've focussed on getting the structural data into a usable format & leaving the metadata reasonably unstructured (it's basically a nested data structure so everything is accessible but not super elegantly). I think there's a lot of good stuff in the metadata compared to PDB files that should eventually be made more accessible through the API.

I've also not managed to get the parser to handle 'broken' files (eg missing elements) especially well so I'll forgo the element type imputation - I don't think that's a major requirement for most people.

I also think adding write functionality can be done down the line (once I've got my head around the format contents). Perhaps we can make a dev release and dogfood/solicit some feedback from the people who have requested the feature? I think the goal is to get it to a usable state (which I think it is now) and the trimmings can come over time. Finishing up the last of the tests now - will try to get to some docs/tutorials next week though the API is more or less identical to the the PDB module.

**Comment by a-r-j** [comment]
Okay, I think this is ready for review! 🥳

**Comment by a-r-j** [comment]
Need to check if this is a concern here: biopython/biopython#3801

**Comment by rasbt** [comment]
That's an interesting concern regarding the optional entries. Do you know if/how they are currently handled in your code?

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [inline biopandas/mmcif/mmcif_parser.py:9]: Shouldn't 

    import json

work on all modern Python versions? I think you maybe added it because you were concerned that it doesn't work in old Python versions. However, I think people who are using older versions have bigger issues due to the type hints in this code :P. So, I think it is safe to assume that this import here is not an issue and the code can be simplified to just

    import json

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [inline biopandas/mmcif/mmcif_parser.py:20]: Do we know the type(s) of exception that would normally occur here?

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [inline biopandas/mmcif/mmcif_parser.py:27]: Shouldn't we use camel (or at least first-letter uppercase) for this to be consistent with Python style?

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [inline biopandas/mmcif/mmcif_parser.py:37]: for methods in Python, the convention is not to use CamelCase but lowercase. E.g., `add_name`

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [reply biopandas/mmcif/mmcif_parser.py:27]: E.g., `_Loop`

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [inline biopandas/mmcif/mmcif_parser.py:38]: I am not quite sure what this is doing to be honest :). 

If `type(name) == str` is True and `partitionString(...)` is not None, then 

    catName = str

and otherwise

    catName = ["", "", ""]

is this intended?

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [inline biopandas/mmcif/mmcif_parser.py:54]: could we change all the methods to the Python function syntax, e.g., `push_value` etc.?

**Review by rasbt — CHANGES_REQUESTED** [review]
Great PR so far! I have a few minor things for my first pass ...

File: biopandas/mmcif/mmcif_parser.py
Comment (a-r-j) [reply biopandas/mmcif/mmcif_parser.py:9]: So, the mmCIF parser implementation comes from PDBJapan. I would think supporting older python versions is their goal too. 

Now, in terms of how we proceed, I think it's your call as the maintainer. AFAIK numpy and pandas have [dropped](https://numpy.org/news/) offical [support](https://pandas.pydata.org/docs/getting_started/install.html) for <3.8 so we could take that as indication to let go of the past 😁 . Do you know if there are a lot of legacy projects that depend on biopandas? Hopefully they'd have pinned a version if so.

Happy to remove the ugly old python support if that's what you'd prefer.

File: biopandas/mmcif/mmcif_parser.py
Comment (a-r-j) [reply biopandas/mmcif/mmcif_parser.py:20]: I expect it would be an `AttributeError`. This was added in Python [2.5 ](https://docs.python.org/2/library/stdtypes.html#str.partition) - probably safe to assume we don't need to support?

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [reply biopandas/mmcif/mmcif_parser.py:9]: I am honestly fine with not supporting Python 2.7 anymore. I think that ship has sailed :P. The reason why I brought this up though is that we have type hints in your code, which are only supported from Python 3.5 onwards anyway. So in this case, code wouldn't be compatible to older Python versions anyways.

File: biopandas/mmcif/mmcif_parser.py
Comment (a-r-j) [reply biopandas/mmcif/mmcif_parser.py:27]: In general, I'd agree. Although In this case, `_loop` is a [directive in the mmCIF format](https://mmcif.wwpdb.org/docs/tutorials/mechanics/pdbx-mmcif-syntax.html) so I think it might actually make it more readable at the expensive of being truly pythonic. I don't have particularly strong feelings either way.

File: biopandas/mmcif/mmcif_parser.py
Comment (a-r-j) [reply biopandas/mmcif/mmcif_parser.py:37]: Agreed 🐫 will fix

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [reply biopandas/mmcif/mmcif_parser.py:9]: PS: I am also fine with not worrying about pre Python 3.8. To be honest, maintaining a library is already a lot of work for a current Python version :)

File: biopandas/mmcif/mmcif_parser.py
Comment (a-r-j) [reply biopandas/mmcif/mmcif_parser.py:54]: Yep! 🐫

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [reply biopandas/mmcif/mmcif_parser.py:20]: Yeah, we can remove the except then.

File: biopandas/mmcif/mmcif_parser.py
Comment (a-r-j) [reply biopandas/mmcif/mmcif_parser.py:9]: Absolutely! I'd be very happy if we could keep support for 3.7 for a while at least. My own project, [Graphein](https://github.com/a-r-j/graphein) depends quite heavily on biopandas and I've had explicit requests to support 3.7. I don't think there's anything here that would prevent that though.

Comment (rasbt) [reply biopandas/mmcif/mmcif_parser.py:9]: Of course, I am happy to not break Python 3.7 support. I do think there is nothing really that should prevent it though. When I have time this weekend, I will try to also shift this project here to GitHub actions and add Python 3.7 as minimum requirement. Graphein looks like a cool library btw. Super impressive.

File: biopandas/mmcif/mmcif_parser.py
Comment (a-r-j) [reply biopandas/mmcif/mmcif_parser.py:38]: It's not a very intuitive line. The behaviour is:

```python
# name: str = "biopandas"
# catName = ('biopandas', '', '')

# name: str = "bio.pandas"
# catName = ('bio', '.', 'pandas')

# name: int = 12
# catName = ['', '', '']
```

File: biopandas/mmcif/mmcif_parser.py
Comment (rasbt) [reply biopandas/mmcif/mmcif_parser.py:27]: Oh I see. So was there an intention to keep this a private class? Because the current code is `_loop` and the format specifies the `loop_` directive.

In any case, I think that `Loop_`, `LoopDir` or `LoopMMCIF` would be better to make sure it's clear that it's a class.

**Comment by rasbt** [comment]
Just had some spare time to take another look and started creating the docs. This is amazing, I was able to run 95% of the PDB tutorial. I think it might be ready to be merged for an early dev release. There are just a few small things:

## 1. 

I renamed `PandasMMCIF` to `PandasMmcif` for consistency with `PandasPdb` (originally, it was `PandasPDB`, but I can't remember why I changed it to `PandasPdb`)

## 2. 
The B-factor calculation is not quite right I think. Might have something to do with reading the file or getting carbon atoms. 


```python
l_1 = PandasMmcif().read_mmcif('./data/lig_conf_1.cif')
l_2 = PandasMmcif().read_mmcif('./data/lig_conf_2.cif')


r = PandasMmcif.rmsd(l_1.df['HETATM'], l_2.df['HETATM'], 
                     s='carbon') # carbon atoms only
print('RMSD: %.4f Angstrom' % r)
```
```
RMSD: 1.7249 Angstrom
```


whereas 

```python
from biopandas.pdb import PandasPdb

l_1 = PandasPdb().read_pdb('./data/lig_conf_1.pdb')
l_2 = PandasPdb().read_pdb('./data/lig_conf_2.pdb')

r = PandasPdb.rmsd(l_1.df['HETATM'], l_2.df['HETATM'], 
                   s='carbon') # carbon atoms only
print('RMSD: %.4f Angstrom' % r)
```

```
RMSD: 3.1405 Angstrom
```

(This is from the tutorial notebooks under docs/tutorials.

## 3. 

In the `mmcif_parser.py` file, we have 2 undefined variables:

1. throw (line 240)
2. unicode (line 288)

**Comment by a-r-j** [comment]
Awesome, thanks for that! Glad it's working out :) 

1. Naming consistency sounds like a good thing!
2. Could this be related to #95? Were the PDB tutorial re-run after this fix? Could explain the discrepancy.
3. Will check out those variables. I noticed them too - I assumed they were keywords from an old Python version but couldn't find anything about them.

**Comment by a-r-j** [comment]
So looking at the code again, I think the `force` is there intentionally to force parsing the file rather than the dumped JSON. It doesn't seem very Pythonic - I guess raising an exception would work but also it's not really an exception but rather a 'creative' way to skip to the `except` block under a particular condition.

**Comment by rasbt** [comment]
> I think this is related to https://github.com/rasbt/biopandas/issues/95. Was the PDB tutorial re-run after this fix? I just re-ran it locally and got RMSD: 1.7249 Angstrom (consistent with the MMCif result)

Oops, you are right 😅

Ok, I think everything looks fine to me then. There are only two things left:

## 1)

I had placeholders for the format explanation method. I don't think we necessarily have to explain it like in the PDB tutorial but could just point readers to the website.
Screen Shot 2022-04-02 at 5 43 50 PM

## 2)
The other thing though is the B-factor sections, I accidentally lost that part while writing the previous comment above. Above, I am referring to B-factor when I show the RMSD issue, but there were two points here. One is the RMSD issue, which solved itself as per your comment :).
The other is the b-factor. I used "occupancy" in the tutorials, but I don't think that represents the b-factor in the mmcif files. Do you know which column the B-factor is stored?
Screen Shot 2022-04-02 at 5 47 20 PM

**Comment by a-r-j** [comment]
I think the correct column is `B_iso_or_equiv`

Another thing that could be nice is a method to convert the mmCIF dataframe into the PDB dataframe format (basically just rename columns & drop any that aren’t present in the PDB) so there can be a degree of inter-operability.

**Comment by rasbt** [comment]
Ah thanks, don't know why I didn't see it. The b-factor sections are fixed now.

- Another thing that could be nice is a method to convert the mmCIF dataframe into the PDB dataframe format (basically just rename columns & drop any that aren’t present in the PDB) so there can be a degree of inter-operability.

Sure, that would be super nice. It doesn't have to be today though and we could add that later. My rough goal for today (Sun) is to have a release some time so people can start using it. And then we could add this as a feature later -- not sure how much work it is (probably easy to implement, but more finicky to test).

Btw. I was just adding some descriptions about the columns to the tutorials and noticed the following:

ATOM & HETATM have the following columns that ANISOU does not have:

```
{'B_iso_or_equiv',
 'Cartn_x',
 'Cartn_y',
 'Cartn_z',
 'auth_asym_id',
 'auth_atom_id',
 'auth_comp_id',
 'auth_seq_id',
 'group_PDB',
 'label_alt_id',
 'label_asym_id',
 'label_atom_id',
 'label_comp_id',
 'label_entity_id',
 'label_seq_id',
 'occupancy',
 'pdbx_PDB_model_num',
 'pdbx_formal_charge'}
```

ANISOU has the following columns that ATOM & HETATM don't have

```
{'U[1][1]',
 'U[1][2]',
 'U[1][3]',
 'U[2][2]',
 'U[2][3]',
 'U[3][3]',
 'pdbx_auth_asym_id',
 'pdbx_auth_atom_id',
 'pdbx_auth_comp_id',
 'pdbx_auth_seq_id',
 'pdbx_label_alt_id',
 'pdbx_label_asym_id',
 'pdbx_label_atom_id',
 'pdbx_label_comp_id',
 'pdbx_label_seq_id'}
```

So, ATOM has `'auth_atom_id'` and ANISOU has  `'pdbx_auth_atom_id'`.
Should those be the same column name, i.e, both either  `'auth_atom_id'` or  `'pdbx_auth_atom_id'`?

I started adding some descriptions.

**Comment by a-r-j** [comment]
So the column names are based on what's specified in the mmCIF format for [ANISOU](https://mmcif.wwpdb.org/docs/pdb_to_pdbx_correspondences.html#ANISOU) and [ATOM/HETATM](https://mmcif.wwpdb.org/docs/pdb_to_pdbx_correspondences.html#ATOMP). I thought about standardising these a bit but ultimately thought staying true to the specification is the best course of action.

**Comment by rasbt** [comment]
Ok that's fair! It's a bit strange that they named it slightly differently, but let's go with that.

**Comment by rasbt** [comment]
Merged it into main for now so people can start using it! I think it's in good enough shape for that. Thanks again for contributing this PR!

**Pull request merged**
